vcs/repository.go

```diff
@@ -27,8 +27,9 @@ type Repository interface {
 	// ErrBranchNotFound if no such branch exists.
 	ResolveBranch(name string) (CommitID, error)
 
-	// Branches returns a list of all branches in the repository.
-	Branches() ([]*Branch, error)
+	// Branches returns a list of all branches in the repository, as well as
+	// the total number of branches.
+	Branches(BranchesOptions) (branches []*Branch, total uint, err error)
```

Comment:
This may not be an issue for branches, but I found that having a `total` return value was a bad idea for commits. That's because it requires all implementations to perform an exact count, which is an expensive operation. I.e., if you are listing the most recent 10 commits, that's quick--but the API contract requires the implementation to iterate through ALL commits to perform an exact count (which can take 2-5 seconds on big repos).

Notice that GitHub doesn't return an exact count, I assume for this very reason. See https://github.com/golang/go/commits/master?page=2 at the bottom, where the page links are just "Older" and "Newer" and not a full page count.

Branches are obviously much fewer in number than commits but the same principle applies.

Possible solutions?

* Mention in the docs that the implementation is not required to perform an exact count if `total >= 50` and that callers should treat any value equal to or above 50 as a lower bound.
* Add an extra return value `exactTotal bool` that is false if the count is a lower bound. (Or make a single return value of `type count struct { n uint; exact bool}`.)
* Just return a `more bool` indicating whether there are any branches beyond those returned.
* Punt for now -- assume the number of branches will be small enough.